rmi/clientsrc/TransImpl/CrashDetection.java

```diff
@@ -0,0 +1,42 @@
+package TransImpl;
+
+import ResImpl.*;
+import java.util.*;
+import java.lang.*;
+
+public class CrashDetection extends Thread
+{
+	TransactionManager host;
+	public static final int sleepTime = 1000; // 1 second in ms
+
+	public CrashDetection(TransactionManager host) {
+	 	this.host = host;
+	 	System.out.println("Crash Detectior Started...");
+	}
+
+	public void run() {
+		while(true) {
+			for (Transaction transaction : host.transactions.values()) {
+ 				if(transaction.status == 0) {
+ 					continue;
+ 				}
+ 				long current = (new Date()).getTime();
+ 				if((current - transaction.getTime()) >= transaction.TIME2LIVE) {
+					try {
+ 						host.abort(transaction.id);
+					} catch (Exception e) {
+						System.out.println(e);
+					}
+ 				}
+ 				try {
+ 					Thread.currentThread().sleep(sleepTime);
+ 				} catch (InterruptedException e) {
+ 					System.out.println("Crash detector failed during sleep");
+ 					System.out.println("Restarting Crash detector...");
+ 					host.startDetector();
```

Comment:
this is the only line changed, from before "host.startDetector(this);"
the "this" argument wasn't needed, causing a crash